src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGetContractsFactory.cs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace NuGet.VisualStudio.Contracts
+{
+    /// <summary>Factory to create contract types</summary>
+    /// <remarks>Trying to be forwards compatible with what C#9 records are going to be</remarks>
+    public static class NuGetContractsFactory
```

Comment:
I really wish we avoided this. 

Roslyn doesn't seem to need it & while Moq can't do it, there are other framework that can mock sealed classes.

Reply:
I need examples of Roslyn not needing factories, because [looking at some code I wrote where I used Roslyn, I needed to use a whole bunch of factory classes](https://github.com/NuGet/Entropy/blob/77b6aade521ebe25b0858f88952fa4becfc30d7e/nuget-sdk-usage/nuget-sdk-usage/Updater/SourceUpdater.cs#L15-L24). I guess I could have used a static string with some c# code, and asked the API to parse the string into objects, but I assume internally they'd still be using factories, and would only achieve being slower.

Additionally, for every property on one of their immutable classes, they have a matching `With{PropertyName}`, which creates a new instance which is a clone, with the one property changed. So, consider the `NuGetInstalledPackage` class, using a Roslyn-like API. Our serice would need to use a factory to create an instance with a package id, but no requested range or version. It would then call the instance's `WithRequestedRange` method which returns a new instance with an id and requested range set, but still null version, and the first instance is eligible for garbage collection. Then call the `WithVersion` or `WithRequestedVersion`, which returns another new instance with all the properties set, and the second instance is eligible for garbage collection. So, creating one object with 3 string properties creates 3 instances of the class, 2 of which are immediately eligible for garbage collection.

I'm happy to implement this type of truely immutable (not read-only) class, but I'm not sure that will be compatible with the future record type language feature. I'm also not sure how it works with RPC serialization and deserialization, but then again I have no idea why the current implementation works with an internal constructor and no custom converters.

Reply:
Just looping back here. 

I'm ok with a factory, I'm ok with a public constructor. 
I think I'm fine as long as the type is not mutable.

I don't have enough details to help make an educated decision about whether C# 9 records would be something we can do without a breaking change.